db/migrate/20250212045125_add_type_source_to_reviewable.rb

```diff
@@ -0,0 +1,44 @@
+# frozen_string_literal: true
+class AddTypeSourceToReviewable < ActiveRecord::Migration[7.2]
+  def change
+    add_column :reviewables, :type_source, :string, null: false, default: "unknown"
+
+    # Migrate existing reviewables from plugins to have a type_source
+    DiscoursePluginRegistry.reviewable_types_lookup.each do |reviewable|
+      Reviewable.where(type: reviewable[:klass].sti_name).update_all(
+        type_source: reviewable[:plugin],
+      )
+    end
+
+    # Migrate existing reviewables from core to have a type_source
+    Reviewable.sti_names.each do |type|
+      if DiscoursePluginRegistry
+           .reviewable_types_lookup
+           .map { |r| r[:klass].sti_name }
+           .include?(type)
+        next
+      end
+
+      Reviewable.where(type: type).update_all(type_source: "core")
+    end
+
+    # Migrate existing reviewables from known plugins to have a type_source
+    known_reviewables = {
+      "discourse-ai" => %w[ReviewableAiChatMessage ReviewableAiPost],
+      "discourse-akismet" => %w[
+        ReviewableAkismetPost
+        ReviewableAkismetPostVotingComment
+        ReviewableAkismetUser
+      ],
+      "discourse-antivirus" => ["ReviewableUpload"],
+      "discourse-category-experts" => ["ReviewableCategoryExpertSuggestion"],
+      "discourse-post-voting" => ["ReviewablePostVotingComment"],
+    }
+
+    known_reviewables.each do |plugin, types|
+      types.each do |type|
+        Reviewable.where(type: type, type_source: "unknown").update_all(type_source: plugin)
+      end
+    end
+  end
+end
```

Comment:
> Migrations that lean on application code tend to be extremely brittle. A few years out a site setting may be removed a user field changed, or even worse, the method name could remain but the semantics change leading to subtle migration breakages.
> 
> Instead, always lean directly on data in the database when making decisions in a migration. For example: query the users table directly or the site_settings table directly using DB.query and DB.exec from [mini_sql](https://github.com/discourse/mini_sql).

This is an excerpt from our internal development lore but we generally do not want to reference application code in migrations. Preferably, everything that needs to change in the database is done directly with SQL statements.

Reply:
I definitely see the purpose of that rule, though I'll need to give more thought to how this could work in practice: some of the data that needs to be inserted doesn't exist in the DB, it can only be generated by application code.

Reply:
My worry here is that the value of `DiscoursePluginRegistry.reviewable_types_lookup ` is tied to the specific version of the plugins that are installed as well as whether the site has the plugin enabled. This means that the migration is no longer idempotent and could make it harder for us to reason about what it does.

Since `DiscoursePluginRegistry.reviewable_types_lookup ` is a new API being introduced, perhaps we can just tackle the known plugins with known reviewables and then add specific migrations to cover the plugins we missed in future migrations.

Reply:
Aye, that's a fair point, it could get tricky much further in the future, there are a lot of moving parts that could change the behaviour of the API.

I'll switch to just doing the known types. 🙂